### PR TITLE
CI: Improve FIH job result assesment

### PR DESCRIPTION
**Pull request by Roolli**
This pull request modifies the FIH CI job to fail in case a successful boot happens.

The CI should fail if a successful boot is achieved by bypassing one or two instructions as it would defeat the purpose of the FIH mechanisms. The relevant workflow has also been modified to disable the auto-cancel feature if one of the configs fail.

relevant issue: #1584

### CODE REVIEW

**Review by davidvincze — APPROVED** [review]

**Review by tamasban — APPROVED** [review]

File: ci/fih_test_docker/utils.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        # The address was successfully skipped by the debugger
+        'SKIPPED': 'Address was skipped',
+        'NO_BOOT': 'System not booted (desired behaviour)',
+        'BOOT': 'System booted (undesired behaviour)'
+}
+
+
+def parse_yaml_file(filepath):
+    with open(filepath) as output_yaml_file:
+        results = yaml.safe_load(output_yaml_file)
```
Comment (utzig) [inline ci/fih_test_docker/utils.py:33]: Not really an output file if you are loading it, right? In a limited scope like this just use `f`.

File: ci/fih_test_docker/validate_output.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+from utils import CATEGORIES, parse_yaml_file
+
+
+def validate_output(test_stats, skip_size, fih_level):
+    if(test_stats[CATEGORIES['BOOT']] > 0
```
Comment (utzig) [inline ci/fih_test_docker/validate_output.py:20]: Add a space between `if` and `(`.

File: ci/fih_test_docker/validate_output.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+def main():
+    parser = argparse.ArgumentParser(description='''Process a FIH test output yaml file,
+     and validate no sucessfull boots have happened''')
+    parser.add_argument('filename', help='yaml file to process')
+    parser.add_argument('skip_size', help='instruction skip size')
+    parser.add_argument('fih_level', nargs="?",
+                        help='fault injection hardening level')
+
+    args = parser.parse_args()
+    (test_stats, _, _) = parse_yaml_file(args.filename)
```
Comment (utzig) [inline ci/fih_test_docker/validate_output.py:34]: No need for parenthesis on the left side, also you could just do:

```python
test_stats = parse_yaml_file(args.filename)[0]
```

File: ci/fih_test_docker/utils.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                continue
+
+            if test["boot"]:
+                test_stats.update([CATEGORIES['BOOT']])
+                continue
+
+            failed_boot_last_lines.update([test["last_line"]])
+        else:
+            exec_fail_reasons.update([test["test_exec_fail_reason"]])
+    return (test_stats, failed_boot_last_lines, exec_fail_reasons)
```
Comment (utzig) [inline ci/fih_test_docker/utils.py:62]: Add on empty line above the return, and the return does not need parenthesis here.

File: ci/fih_test_docker/generate_test_report.py [hunk 5]
```diff
@@ -13,47 +13,12 @@
...
-                continue
-
-            if test["boot"]:
-                test_stats.update([CATEGORIES['BOOT']])
-                continue
-
-            failed_boot_last_lines.update([test["last_line"]])
-        else:
-            exec_fail_reasons.update([test["test_exec_fail_reason"]])
+    (test_stats, failed_boot_last_lines, exec_fail_reasons) = results
```
Comment (utzig) [inline ci/fih_test_docker/generate_test_report.py:21]: No need for parenthesis.

File: .github/workflows/fih_tests.yaml [hunk 6]
```diff
@@ -13,13 +13,17 @@ concurrency:
 jobs:
   config:
     strategy:
+      fail-fast: false
```
Comment (utzig) [inline .github/workflows/fih_tests.yaml:16]: Why?

File: ci/fih_test_docker/validate_output.py (hunk 2 above)
Comment (Roolli) [reply ci/fih_test_docker/validate_output.py]: Fixed

File: ci/fih_test_docker/validate_output.py (hunk 3 above)
Comment (Roolli) [reply ci/fih_test_docker/validate_output.py]: Fixed

File: ci/fih_test_docker/utils.py (hunk 4 above)
Comment (Roolli) [reply ci/fih_test_docker/utils.py]: Fixed

File: ci/fih_test_docker/generate_test_report.py (hunk 5 above)
Comment (Roolli) [reply ci/fih_test_docker/generate_test_report.py]: Fixed

File: .github/workflows/fih_tests.yaml (hunk 6 above)
Comment (Roolli) [reply .github/workflows/fih_tests.yaml:16]: This reduces the amount of times you need to run the whole CI, if there are different errors between different configs, but @tamasban can clarify it more.

Comment (utzig) [reply .github/workflows/fih_tests.yaml:16]: I have never used this option, but I think that failing fast is better than failing slow! :rofl:

Comment (utzig) [reply .github/workflows/fih_tests.yaml:16]: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

Maybe what you want to achieve is to let other workflows run to the end even if the FIH tests fail? @tamasban

Comment (tamasban) [reply .github/workflows/fih_tests.yaml:16]: AFAIU this config option has no effect on other workflows like sim, Zephyr, etc. Its scope is restricted only to the FIH jobs.

The goal is to let the other jobs run until the end within the FIH test suite even though one fails. With this approach, one can see all build or FI-related issues after a single CI execution and failures are not revealed only one by one in consecutive CI executions.

This is not aligned with the rest of the CI job configs. We can revert this change if you want to keep all the jobs config aligned.

**Review by utzig — APPROVED** [review]
LGTM

**Review by d3zd3z — APPROVED** [review]

**Pull request merged**